filebeat/input/filestream/input_test.go

```diff
@@ -0,0 +1,332 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/statestore"
+	"github.com/elastic/beats/v7/libbeat/statestore/backend/memlog"
+)
+
+// test_close_renamed from test_harvester.py
+func TestFilestreamCloseRenamed(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFilestreamCloseRenamed")
+	if err != nil {
+		t.Fatalf("cannot create temporary folder for test file: %+v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	testlogPath := filepath.Join(tmpDir, "test.log")
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	testStore := openTestStatestore(t, tmpDir)
+	inp := getInputFromConfig(
+		t,
+		testStore,
+		map[string]interface{}{
+			"paths":                                []string{testlogPath},
+			"prospector.scanner.check_interval":    "1s",
+			"close.on_state_change.check_interval": "1s",
+			"close.on_state_change.renamed":        "true",
+		},
+	)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	inputCtx := input.Context{Logger: logp.L(), Cancelation: ctx}
+	pipeline := &mockPipelineConnector{}
+
+	var wg sync.WaitGroup
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+
+		inp.Run(inputCtx, pipeline)
+	}()
+
+	// first event has made it successfully
+	waitUntilEventCount(1, pipeline)
+	// check registry
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+
+	testlogPathRenamed := filepath.Join(tmpDir, "test.log.rotated")
+	err = os.Rename(testlogPath, testlogPathRenamed)
+	if err != nil {
+		t.Fatalf("failed to rename file '%s': %+v", testlogPath, err)
+	}
+
+	// wait enough time so renaming of file can be detected
+	time.Sleep(2 * time.Second)
+
+	err = ioutil.WriteFile(testlogPath, []byte("new first log line\nnew second log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	// new two events are read
+	waitUntilEventCount(2, pipeline)
```

Comment:
We sent 3 events in total. When the harvester first starts, it connects with the pipeline and a `beat.Client` is created. This client gets the first event and then as the file was renamed and `reader.on_state_change.renamed` is enabled, the `Harvester` is stopped and the `beat.Client` instance is closed. When two new lines show up in the newly created file, a new `beat.Client` is created and a new `Harvester` is started for reading. Thus, we get 1 event in the first round and then 2 more later.

Reply:
Thanks for the explanation. My question is why use `waitUntilEventCount(2, pipeline)` instead of `waitUntilEventCount(3, pipeline)`. I guess 2 is good enough to learn that the second file started to be collected?